core/src/test/scala/org/apache/spark/ConsistentAccumulatorsSuite.scala

```diff
@@ -0,0 +1,284 @@
+        val List(racc, racc1, racc2, racc3) = raccs
+        val c = input.combineByKey(
+          (x: Int) => {acc1 += 1; acc += 1; racc1 += 1; racc += 1; x},
+          {(a: Int, b: Int) => acc2 += 1; acc += 1; racc2 += 1; racc += 1; (a + b)},
+          {(a: Int, b: Int) => acc3 += 1; acc += 1; racc3 += 1; racc += 1; (a + b)},
+          new HashPartitioner(10),
+          mapSideCombine)
+        val d = input.combineByKey(
+          (x: Int) => {acc1 += 1; acc += 1; x},
+          {(a: Int, b: Int) => acc2 += 1; acc += 1; (a + b)},
+          {(a: Int, b: Int) => acc3 += 1; acc += 1; (a + b)},
+          new HashPartitioner(2),
+          mapSideCombine)
+        val e = d.map{x => acc += 1; x}
+        c.count()
+        // If our partitioner is known then we should only create
+        // one combiner for each key value. Otherwise we should
+        // create at least that many combiners.
+        if (input.partitioner.isDefined) {
+          acc1.value should be (buckets)
+        } else {
+          acc1.value should be >= (buckets)
+        }
+        if (input.partitioner.isDefined) {
+          acc2.value should be > (0)
+        } else if (mapSideCombine) {
+          acc3.value should be > (0)
+        } else {
+          acc2.value should be > (0)
+          acc3.value should be (0)
+        }
+        acc.value should be (acc1.value + acc2.value + acc3.value)
+        val oldValues = accs.map(_.value)
+        // For one action the consistent accumulators and regular should have the same value.
+        accs.map(_.value) should be (raccs.map(_.value))
+        c.count()
+        accs.map(_.value) should be (oldValues)
+        // Executing a second _different_ aggregation should count everything 2x
+        d.count()
+        accs.map(_.value) should be (oldValues.map(_ * 2))
+        // Computing the mapped value on top and verify new changes are processed and old changes
+        // are note double counted.
+        val count = e.count()
+        accs.tail.map(_.value) should be (oldValues.tail.map(_ * 2))
+        acc.value should be (oldValues.head * 2 + count)
+      }
+    }
+  }
+
+  test("map + cache + first + count") {
+    sc = new SparkContext("local[2]", "test")
+    val acc : Accumulator[Int] = sc.accumulator(0, consistent = true)
+
+    val a = sc.parallelize(1 to 20, 10)
+    val b = a.map{x => acc += x; x}
+    b.cache()
+    b.first()
+    acc.value should be > (0)
+    b.collect()
+    acc.value should be (210)
+  }
+
+  test("coalesce acc on either side") {
+    sc = new SparkContext("local[2]", "test")
+    val List(acc1, acc2, acc3) = 1.to(3).map(x => sc.accumulator(0, consistent = true)).toList
+    val a = sc.parallelize(1 to 20, 10)
+    val b = a.map{x => acc1 += x; acc2 += x; 2 * x}
+    val c = b.coalesce(2).map{x => acc1 += x; acc3 += x; x}
+    c.count()
+    acc1.value should be (630)
+    acc2.value should be (210)
+    acc3.value should be (420)
+  }
```

Comment:
I think the coalesce test needs to do partial partition reading, of both the original partitions and the coalesced partitions, like so (this passes):

```scala
  test("coalesce with partial partition reading") {
    // when coalescing, one task can completely read partitions of the input RDDs while not reading complete
    // complete partitions of the post-coalesce RDD.  We make sure that the accumulator has consistent semantics
    // in these cases.
    sc = new SparkContext("local[2]", "test")
    val List(acc1, acc2, acc3) = 1.to(3).map(x => sc.accumulator(0, consistent = true)).toList
    val a = sc.parallelize(1 to 20, 10)
    val b = a.map{x => acc1 += x; acc2 += x; 2 * x}
    val c = b.coalesce(2).map{x => acc1 += x; acc3 += x; x}
    // we read all of partition 1 from RDD's a & b, and part of partition 2
    // however, for RDD c, we don't read any of the partitions fully
    // so we should get updates for 1 partition from b, and nothing from c
    c.take(3) should be ((1 to 3).map(_*2).toArray)
    acc1.value should be (3) // only elements 1 & 2 came from fully read partitions
    acc2.value should be (3)
    acc3.value should be (0)

    // now we read a few more parts:
    c.take(9) should be ((1 to 9).map(_*2).toArray)
    acc1.value should be (36)
    acc2.value should be (36)
    acc3.value should be (0)


    // a couple more, this time we read 1 of c's partitions fully
    c.take(15) should be ((1 to 15).map(_*2).toArray)
    acc1.value should be (215)
    acc2.value should be (105)
    acc3.value should be (110)

    // and if we read the entire data, we get the entire value
    c.count()
    acc1.value should be (630)
    acc2.value should be (210)
    acc3.value should be (420)
  }
```